bee-network/bee-autopeering/src/discovery/manager.rs

```diff
@@ -0,0 +1,1000 @@
+                // Panic: unwrapping is fine, because we checked that the list isn't empty, and `index` must be in
+                // range.
+                let peer = replacements.write().remove_at(index).unwrap();
+
+                active_peers.insert(peer.into());
+            }
+        }
+    }
+}
+
+pub(crate) struct RecvContext<'a> {
+    peer_id: &'a PeerId,
+    msg_bytes: &'a [u8],
+    server_tx: &'a ServerTx,
+    local: &'a Local,
+    request_mngr: &'a RequestManager,
+    peer_addr: SocketAddr,
+    event_tx: &'a EventTx,
+    active_peers: &'a ActivePeersList,
+    replacements: &'a ReplacementList,
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+// VALIDATION
+///////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub(crate) enum ValidationError {
+    // The protocol version must match.
+    #[error("version mismatch; expected: {expected}, received: {received}")]
+    VersionMismatch { expected: u32, received: u32 },
+    // The network id must match.
+    #[error("network id mismatch; expected: {expected}, received: {received}")]
+    NetworkIdMismatch { expected: u32, received: u32 },
+    // The request must not be expired.
+    #[error("request expired")]
+    RequestExpired,
+    // The response must arrive in time.
+    #[error("no corresponding request, or timeout")]
+    NoCorrespondingRequestOrTimeout,
+    // The hash of the corresponding request must be correct.
+    #[error("incorrect request hash")]
+    IncorrectRequestHash,
+    // The peer must have an autopeering service.
+    #[error("no autopeering service")]
+    NoAutopeeringService,
+    // The service port must match with the detected port.
+    #[error("service port mismatch; expected: {expected}, received: {received}")]
+    ServicePortMismatch {
+        expected: ServicePort,
+        received: ServicePort,
+    },
+}
+
+fn validate_verification_request(
+    verif_req: &VerificationRequest,
+    version: u32,
+    network_id: u32,
+) -> Result<(), ValidationError> {
+    use ValidationError::*;
+
+    if verif_req.version() != version {
+        Err(VersionMismatch {
+            expected: version,
+            received: verif_req.version(),
+        })
+    } else if verif_req.network_id() != network_id {
+        Err(NetworkIdMismatch {
+            expected: network_id,
+            received: verif_req.network_id(),
+        })
+    } else if request::is_expired(verif_req.timestamp()) {
+        Err(RequestExpired)
+    } else {
+        // NOTE: the validity of the transmitted source and target addresses is ensured through the
+        // `VerificationRequest` type.
+        // TODO: maybe add check whether the peer sent the correct source address in the packet.
+        // TODO: store own external IP address as perceived by the peer
```

Comment:
Reminder

Reply:
Thanks, but aren't TODOs reminders? Before releases I will `grep` for TODOs to check whether I should solve any of those. I could remove all those and put them in a separate file locally to not have them in the codebase, but those TODOs can also be something interested community devs could work on if they wanted to to get a hang of the codebase, so I personally like to leave them in there.